pil-analyzer/src/expression_processor.rs

```diff
@@ -41,51 +41,57 @@ impl<'a, D: AnalysisDriver> ExpressionProcessor<'a, D> {
     pub fn process_array_expression(
         &mut self,
         array_expression: ::powdr_ast::parsed::ArrayExpression,
-    ) -> ArrayExpression<Reference> {
+    ) -> Result<ArrayExpression<Reference>, Error> {
         match array_expression {
             ArrayExpression::Value(expressions) => {
-                let values = self.process_expressions(expressions);
-                ArrayExpression::Value(values)
+                let values = self.process_expressions(expressions)?;
+                Ok(ArrayExpression::Value(values))
             }
-            ArrayExpression::RepeatedValue(expressions) => {
-                ArrayExpression::RepeatedValue(self.process_expressions(expressions))
-            }
-            ArrayExpression::Concat(left, right) => ArrayExpression::Concat(
-                Box::new(self.process_array_expression(*left)),
-                Box::new(self.process_array_expression(*right)),
-            ),
+            ArrayExpression::RepeatedValue(expressions) => Ok(ArrayExpression::RepeatedValue(
+                self.process_expressions(expressions)?,
+            )),
+            ArrayExpression::Concat(left, right) => Ok(ArrayExpression::Concat(
+                Box::new(self.process_array_expression(*left)?),
+                Box::new(self.process_array_expression(*right)?),
+            )),
         }
     }
 
-    pub fn process_expressions(&mut self, exprs: Vec<parsed::Expression>) -> Vec<Expression> {
+    pub fn process_expressions(
+        &mut self,
+        exprs: Vec<parsed::Expression>,
+    ) -> Result<Vec<Expression>, Error> {
         exprs
             .into_iter()
             .map(|e| self.process_expression(e))
-            .collect()
+            .collect::<Result<Vec<_>, _>>()
     }
 
-    pub fn process_expression(&mut self, expr: parsed::Expression) -> Expression {
+    pub fn process_expression(&mut self, expr: parsed::Expression) -> Result<Expression, Error> {
         use parsed::Expression as PExpression;
         match expr {
```

Comment:
I often use the style of `Ok(match expr{...`, then you don't need an `Ok(` in every single branch.